test/cpp/end2end/xds/xds_security_end2end_test.cc

```diff
@@ -782,6 +784,350 @@ TEST_P(XdsSecurityTest, MtlsWithAggregateCluster) {
+            upstream_tls_context.mutable_common_tls_context()
+                ->mutable_validation_context();
+        for (const auto& san_matcher : san_matchers) {
+          *validation_context->add_match_subject_alt_names() = san_matcher;
+        }
+      }
+      upstream_tls_context.set_sni(std::string(sni));
+      upstream_tls_context.set_auto_host_sni(auto_host_sni);
+      upstream_tls_context.set_auto_sni_san_validation(auto_sni_san_validation);
+      transport_socket->mutable_typed_config()->PackFrom(upstream_tls_context);
+    }
+  }
+  static const char kSni1CertPath[];
+  static const char kSni1KeyPath[];
+  static const char kSni2CertPath[];
+  static const char kSni2KeyPath[];
+  static const char kSni3CertPath[];
+  static const char kSni3KeyPath[];
+
+  std::string root_cert_;
+  grpc_core::PemKeyCertPairList identity_pair_;
+  grpc_core::RefCountedPtr<grpc_core::FakeResolverResponseGenerator>
+      logical_dns_cluster_resolver_response_generator_;
+};
+
+static grpc_core::EndpointAddressesList CreateAddressListFromPortList(
+    const std::vector<int>& ports) {
+  grpc_core::EndpointAddressesList addresses;
+  for (int port : ports) {
+    absl::StatusOr<grpc_core::URI> lb_uri =
+        grpc_core::URI::Parse(grpc_core::LocalIpUri(port));
+    GRPC_CHECK_OK(lb_uri);
+    grpc_resolved_address address;
+    GRPC_CHECK(grpc_parse_uri(*lb_uri, &address));
+    addresses.emplace_back(address, grpc_core::ChannelArgs());
+  }
+  return addresses;
+}
+
+const char XdsSniSecurityTest::kSni1CertPath[] =
+    "src/core/tsi/test_creds/sni1.pem";
+const char XdsSniSecurityTest::kSni1KeyPath[] =
+    "src/core/tsi/test_creds/sni1.key";
+const char XdsSniSecurityTest::kSni2CertPath[] =
+    "src/core/tsi/test_creds/sni2.pem";
+const char XdsSniSecurityTest::kSni2KeyPath[] =
+    "src/core/tsi/test_creds/sni2.key";
+const char XdsSniSecurityTest::kSni3CertPath[] =
+    "src/core/tsi/test_creds/sni3.pem";
+const char XdsSniSecurityTest::kSni3KeyPath[] =
+    "src/core/tsi/test_creds/sni3.key";
+
+INSTANTIATE_TEST_SUITE_P(XdsTest, XdsSniSecurityTest,
+                         ::testing::Values(XdsTestType()), &XdsTestType::Name);
+
+TEST_P(XdsSniSecurityTest, TestLegacySniBehavior) {
+  grpc_core::testing::ScopedExperimentalEnvVar env(
+      "GRPC_USE_CHANNEL_AUTHORITY_IF_NO_SNI_APPLICABLE");
+  g_fake1_cert_data_map->Set({{"", {root_cert_, identity_pair_}}});
+  const char* kEdsServiceName = "eds_service_name";
+  EdsResourceArgs args({
+      {"locality0", CreateEndpointsForBackends(0, 1)},
+  });
+  balancer_->ads_service()->SetEdsResource(
+      BuildEdsResource(args, kEdsServiceName));
+  auto cluster = default_cluster_;
+  cluster.mutable_eds_cluster_config()->set_service_name(kEdsServiceName);
+  StringMatcher san_matcher;
+  // Expect the default channel target (server.example.com)
+  san_matcher.set_exact("server.example.com");
+  MaybeSetUpstreamTlsContextWithSniOnCluster("fake_plugin1", "", "", false,
```

Comment:
For each of the string and boolean parameters, suggest adding a comment like `/*cert_provider_name=*/` right before each parameter, so that it's clearer to the reader what's actually being configured.

Same for all tests.